### PR TITLE
fix(automodel): plumb packing boundaries as top-level cu_seqlens for Nemotron

### LINKED ISSUE #4167: [Bug] Packed Nemotron sequences attend across trajectory boundaries in FSDP2/AutoModel
## Goal

Match the completed PrimeRL Example 5 setup as closely as supported by NeMo RL + NeMo Gym: full-parameter, text-only Nemotron 3.5 Super GRPO using the same Harbor SimpleQA task split and grader.

| Setting | Completed PrimeRL Example 5 | NeMo attempt |
| --- | --- | --- |
| GPU placement | 16 training + 16 inference GPUs | Matched |
| Agent / task | Verifiers Bash + Harbor SimpleQA | Terminus2NemoGym + the same Harbor task split and grader |
| Parameter / optimizer CPU offload | Both off | Matched |
| Activation CPU offload | On, alongside checkpointing | No equivalent exposed in this AutoModel path; checkpointing on |
| Scheduling | Async; maximum policy age 1 | Async; maximum trajectory age 1; implementation differs |
| Batch / group size | 256 / 8 | Matched target; no dynamic replacement |
| Training parallelism | FSDP2 across 16 training GPUs: DP shard 16, DP replicate 1; TP / CP / EP = 1 / 1 / 1 | Matched |
| Precision | BF16 compute; FP32 stored parameters and optimization | FP32 stored parameters; initially FP32 moments, then BF16 retries |
| LR / betas / weight decay / grad norm cap | 1e-5 / 0.9, 0.999 / 0 / 1 | Matched |
| Context / output / maximum turns | 32K / 4K / 8 | Matched limits; actual conversations differ |
| Inference parallelism | 4 separate vLLM engines × TP4 = 16 GPUs; EP1, CUDA graphs, 0.85 memory fraction | Matched settings; different vLLM versions |
| Output-layer memory | Fused, chunked forward and backward | Full FP32 logits and training log-softmax |
| Validation | 16 before/after; temperature 0; interval 10 | Same settings; failed before final validation |
| Agent / verifier timeout | 600 / 120 seconds | Matched |
| Rollout concurrency | Adaptive 8–128, starting at 64 | Fixed maximum 128 |

FSDP2 shards parameters, gradients and optimizer state across the 16 training
ranks; each rank processes different trajectories. TP/CP/EP = 1 therefore still
uses distributed training, without tensor, context or expert parallelism.

## Environment

- NeMo RL: `6453bfd66bda1572654b679fb7a9194f3033b148` for the source paths below.
- AutoModel 0.7.0, submodule `6ae5ca23895d1e06c81cc6103732a20510833fda`; PyTorch 2.11.0+cu130; Transformers 5.15.1.
- Training workload: eight GB200 nodes, four GPUs/node; 16 training GPUs and 16 inference GPUs.
- NeMo Gym: `c3bac96314a59f28b896f597eb9845d175bb0252`; Terminus2NemoGym, maximum 8 turns, summarization disabled.

## Description

The NeMo RL packing path passes sequence boundaries in nested `flash_attn_kwargs`, but the native Nemotron attention backend reads top-level boundary arguments. In a CUDA component probe, changing the first trajectory changes the second trajectory's output despite supplying NeMo RL's packing metadata.

This breaks the isolation required for training independent packed trajectories. The equivalent native boundary/layout interface passes the isolation control.

## Steps to reproduce

1. Use the NeMo RL and AutoModel revisions above with CUDA, BF16 and Transformer Engine available. The probe uses random weights and needs neither checkpoint weights nor Harbor services.
2. Save the script below as `check_packing_boundaries.py` and run it with the AutoModel worker interpreter: `python-DTensorPolicyWorkerV2 check_packing_boundaries.py` in the retained runtime, or `python check_packing_boundaries.py` in an equivalent installed environment.
3. The probe creates two 64-token trajectories, changes only the first, and compares the second trajectory's output. It repeats the check using native top-level `cu_seqlens` and THD layout (flattened tokens rather than batch × sequence).

<details>
<summary>Standalone component reproducer</summary>

```python
"""Lyris-only diagnostic: does NeMo RL packing metadata isolate TE attention?"""

import json
from types import SimpleNamespace

import torch
from nemo_automodel.components.models.common import BackendConfig
from nemo_automodel.components.models.nemotron_v3.layers import NemotronV3Attention
from nemo_rl.models.huggingface.common import get_flash_attention_kwargs


def main() -> None:
    torch.manual_seed(42)
    assert torch.cuda.is_available()
    cfg = SimpleNamespace(
        hidden_size=512,
        num_attention_heads=8,
        num_key_value_heads=4,
        head_dim=64,
        torch_dtype="bfloat16",
        attention_dropout=0.0,
    )
    layer = (
        NemotronV3Attention(cfg, BackendConfig(attn="te", linear="torch")).cuda().eval()
    )
    length = 64
    first = torch.randn(1, length, 512, device="cuda", dtype=torch.bfloat16)
    second = torch.randn_like(first)
    changed_first = torch.randn_like(first)
    flash_kwargs = get_flash_attention_kwargs(
        torch.tensor([length, length], device="cuda")
    )
    bounds = torch.tensor([0, length, 2 * length], device="cuda", dtype=torch.int32)
    with torch.no_grad():
        packed = torch.cat((first, second), dim=1)
        changed = torch.cat((changed_first, second), dim=1)
        # Exact nested metadata key passed by NeMo RL _build_model_batch.
        actual = layer(packed, flash_attn_kwargs=flash_kwargs)[:, length:]
        perturbed = layer(changed, flash_attn_kwargs=flash_kwargs)[:, length:]
        # Positive control: the same module with its native THD boundary interface.
        native = layer(packed.squeeze(0), cu_seqlens=bounds, max_seqlen=length)[length:]
        native_changed = layer(
            changed.squeeze(0), cu_seqlens=bounds, max_seqlen=length
        )[length:]
        standalone = layer(second)
    result = {
        "scope": "one native NemotronV3 TE attention layer; random weights; CUDA BF16",
        "nemo_metadata_second_sequence_max_delta": (actual - perturbed)
        .abs()
        .max()
        .item(),
        "native_metadata_second_sequence_max_delta": (native - native_changed)
        .abs()
        .max()
        .item(),
        "nemo_metadata_vs_standalone_max_delta": (actual - standalone)
        .abs()
        .max()
        .item(),
    }
    print(json.dumps(result, indent=2))
    assert result["native_metadata_second_sequence_max_delta"] == 0, (
        "Positive control failed"
    )
    if result["nemo_metadata_second_sequence_max_delta"] != 0:
        raise SystemExit(
            "FAIL: changing the first trajectory changes the second trajectory"
        )


if __name__ == "__main__":
    main()
```

</details>

## Actual result

Recorded on September 17:

```text
nemo_metadata_second_sequence_max_delta: 0.2177734375
native_metadata_second_sequence_max_delta: 0.0
FAIL: changing the first trajectory changes the second trajectory
```

The exact nonzero delta can vary with the runtime; the failure is any dependency of trajectory B on trajectory A. The full training model loaded with packing enabled, but we stopped before update 1 after this check failed: **0/3 optimizer updates**, not observed corrupted training updates.

## Cause and expected behavior

- [NeMo packing](https://github.com/NVIDIA-NeMo/RL/blob/6453bfd66bda1572654b679fb7a9194f3033b148/nemo_rl/models/automodel/data.py#L262-L279) concatenates trajectories, removes the attention mask, and creates FlashAttention metadata.
- [Model-batch construction](https://github.com/NVIDIA-NeMo/RL/blob/6453bfd66bda1572654b679fb7a9194f3033b148/nemo_rl/models/automodel/train.py#L103-L104) places it under `flash_attn_kwargs`.
- [AutoModel's boundary parser](https://github.com/NVIDIA-NeMo/Automodel/blob/6ae5ca23895d1e06c81cc6103732a20510833fda/nemo_automodel/components/attention/utils.py#L131-L188) consumes top-level `cu_seqlens` or `cu_seqlens_q`/`cu_seqlens_kv`; it does not unpack that nested object.

Translate packing metadata and tensor layout into the native model's supported interface, or reject packing for this backend until it is supported. Validate both attention and recurrent-state isolation: [Nemotron's Mamba path](https://github.com/NVIDIA-NeMo/Automodel/blob/6ae5ca23895d1e06c81cc6103732a20510833fda/nemo_automodel/components/models/nemotron_v3/layers.py#L340-L375) also expects `seq_idx` or top-level `cu_seqlens`.

The measured reproducer covers **one random-weight attention layer**, not Mamba, backward, FSDP2 execution or full-checkpoint equivalence. The native positive control changes both metadata placement and tensor layout. A fix should include packed-versus-separate forward/gradient checks for the hybrid model.

PrimeRL Example 5 completed three updates using its own packing path. We did not run this same component probe against PrimeRL; its success does not by itself establish cross-framework numerical equivalence.

## Run references

- [Retained findings and reproduction guide (NVIDIA access)](https://gitlab-master.nvidia.com/xiyu/benchling-buidathon-super-nemorl/-/blob/xinyu-dev/super35-nemorl/examples/lyris/super35/README.md). Failed-run configs and raw logs were removed during cleanup; the guide preserves their recorded findings.
- [PrimeRL completed-run config](https://github.com/xinyu-dev/benchling-buidathon-super-primerl/blob/222b361ac85c708fbdddd27ba0011fbfc4026697/outputs/nemotron35/grpo-r3-harbor-smoke/configs/resolved/trainer.json) and [results](https://github.com/xinyu-dev/benchling-buidathon-super-primerl/blob/222b361ac85c708fbdddd27ba0011fbfc4026697/outputs/nemotron35/grpo-r3-harbor-smoke/summary.json) (repository access required).

### PR DESCRIPTION
**Pull request by dundysm**
## Summary

Fixes #4167.

Packed Nemotron sequences were attending across trajectory boundaries because NeMo RL packing put boundaries under nested `flash_attn_kwargs` (HF FA2 convention), while Automodel Nemotron TE attention only reads top-level `cu_seqlens` / `qkv_format` via `preprocess_args_and_kwargs_for_attn`. Nested kwargs were ignored, so a packed batch was treated as one causal sequence.

### Approach

- Keep packing itself unchanged (still builds nested `flash_attn_kwargs` in `data.py`).
- In `_build_model_batch`, when the model expects the native packing contract, promote nested metadata to top-level:
  - `cu_seqlens` / `cu_seqlens_q` / `cu_seqlens_kv` (needed for TE attention and Mamba `seq_idx`)
  - `max_seqlen*`
  - `qkv_format="thd"`
- Detection prefers `inspect.signature(model.forward)` for explicit `cu_seqlens` / `cu_seqlens_q` / `cu_seqlens_kv`, then Nemotron `model_type` / class markers. HF FA2 models that only document nested `flash_attn_kwargs` keep the nested path.
- Token layout stays `[1, T]`; Automodel Nemotron squeezes to THD when `qkv_format == "thd"` and TE is active. Gemma3 / multimodal paths that already pop `flash_attn_kwargs` are untouched.

Diff is localized to packing translation helpers plus a small `_build_model_batch` hook to limit conflict risk with #3932 (CPU sequence boundaries perf).

## Test plan

- [x] Unit tests for mapping / detection / promotion (`tests/unit/models/automodel/test_automodel_packing.py`): 6/6 passed locally (CPU, no Automodel runtime).
- [x] `_build_model_batch` cases for native `cu_seqlens` model vs HF FA2-only model (marked `@pytest.mark.automodel`; needs Automodel extra in CI).
- [x] GPU probe adapted from the issue (`test_packing_boundaries_probe.py`) is present and skips without CUDA + TE + Automodel Nemotron attention. Not run here (no CUDA/TE on this box).

## Risks

- Models with `**kwargs` that are not Nemotron and do not declare `cu_seqlens*` keep nested kwargs (conservative default).
- If a future Automodel model expects a different top-level packing key set, detection / promotion may need extending.

@xinyu-dev @yuki-97 @jQizhang

### CODE REVIEW

**Comment by dundysm** [comment]
@xinyu-dev @yuki-97 @jQizhang packing boundary fix for #4167: promote nested flash_attn_kwargs to top-level cu_seqlens / qkv_format=thd for automodel nemotron te (and any forward that accepts cu_seqlens*), leave nested kwargs for hf fa2-only models. unit packing tests green on cpu; gpu probe skipped here (no cuda/te). happy to adjust detection if you prefer a narrower nemotron-only gate.
